src/ucs/arch/x86_64/cpu.c

```diff
@@ -333,4 +469,100 @@ void ucs_cpu_init()
+        ucs_x86_cpuid(X86_CPUID_GET_CACHE_INFO,
+                      ucs_unaligned_ptr(&reg.eax), ucs_unaligned_ptr(&reg.ebx),
+                      ucs_unaligned_ptr(&reg.ecx), ucs_unaligned_ptr(&reg.edx));
+
+        if (i == 0) { /* on first iteration get max iteration number */
+            max_iter     = reg.max_iter;
+            reg.max_iter = 0; /* mask iteration register from processing */
+        }
+
+        for (r = 0; r < ucs_array_size(reg.reg); r++) {
+            if (ucs_test_all_flags(reg.reg[r].value, X86_CPU_CACHE_RESERVED)) {
+                continue;
+            }
+
+            for (t = 0; (t < ucs_array_size(reg.reg[r].tag)) && (reg.reg[r].tag[t] != 0); t++) {
+                tag = reg.reg[r].tag[t];
+
+                switch(tag) {
+                case X86_CPU_CACHE_TAG_L1_ONLY:
+                    level1_only = 1;
+                    break;
+                case X86_CPU_CACHE_TAG_LEAF4:
+                    for (l4 = 0;; l4++) {
+                        ucs_x86_cpuid_ecx(X86_CPUID_GET_LEAF4_INFO, l4,
+                                          ucs_unaligned_ptr(&cache_info.reg),
+                                          ucs_unaligned_ptr(&line_info.reg),
+                                          &sets, ucs_unaligned_ptr(&reg.edx));
+
+                        if (cache_info.type == 0) {
+                            /* we are done - nothing found, go to next register */
+                            break;
+                        }
+
+                        for (c = 0; c < length; c++) {
+                            if ((cache_info.level == x86_cpu_cache[c].level) &&
+                                (cache_info.type  == x86_cpu_cache[c].type)) {
+                                /* found it */
+                                /* cache entry is not updated yet */
+                                /* and cache level is 1 or all levels are supported */
+                                if (!((cache[c] == 0) &&
+                                      ((x86_cpu_cache[c].level == 1) || !level1_only))) {
+                                    break;
+                                }
+
+                                cache[c] = (line_info.associativity + 1) *
+                                           (line_info.partitions + 1)    *
+                                           (line_info.size + 1)          *
+                                           (sets + 1);
+                            }
+                        }
+                    }
+                    return UCS_OK;
+                default:
+                    if ((tag < ucs_array_size(ucs_x86_cpu_known_cache)) && /* tag in table */
+                        (ucs_x86_cpu_known_cache[tag].size != 0) && /* table entry is not empty */
+                        (cache[ucs_x86_cpu_known_cache[tag].type] == 0)) { /* cache is not filled yet */
+                        cache[ucs_x86_cpu_known_cache[tag].type] = ucs_x86_cpu_known_cache[tag].size;
```

Comment:
line too long
maybe split to functions?

Reply:
split into few conditions, added helper variable to simplify reading